Allow chosing Python 2 or Python 3

oxinabox opened the pull request:
I wished to test one of the packages I contribute to on both Python 3 and Python 2.
https://github.com/malmaud/TensorFlow.jl/issues/251
Because I want to use Conda.jl to install my our dependencies, and the python itself,
I needed Conda to download the right versions of the dependencies.
After looking into it, and some discussion with @Ismael-VC 
concluded the most sensible way to do this was to use an environment variable to change the version of Miniconda being used -- thus changing the default python version.

I added Travis Tests, but not AppVeyor tests,
because A. I know Travis matrixes better,
and B. Having one without the environment variables set means that that can be checked also

oxinabox commented:
Failure on Travis seems to be unrelated.

isuruf commented:
What do you think about doing `Conda.add("python=3")` after miniconda is installed ?

oxinabox commented:
>What do you think about doing Conda.add("python=3") after miniconda is installed ?

That was kinda my initial plan, more or less.
But I found that that kind of thing doesn't work for the initial root enviroment.
Because by the time you are able to run `Conda.add`,
it has already downloaded and installed a lot of Python 2 packages.
So you get conflicts:

```julia
julia> using Conda

julia> Conda.add("python=3")
INFO: Downloading miniconda installer ...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 27.7M  100 27.7M    0     0  46.2M      0 --:--:-- --:--:-- --:--:-- 46.2M
INFO: Installing miniconda ...
PREFIX=/home/wheel/oxinabox/.julia/v0.5/Conda/deps/usr
installing: python-2.7.13-0 ...
installing: asn1crypto-0.22.0-py27_0 ...
installing: cffi-1.10.0-py27_0 ...
installing: conda-env-2.6.0-0 ...

...
UnsatisfiableError: The following specifications were found to be in conflict:
  - enum34
  - packaging
  - python 3*
Use "conda info <package>" to see the dependencies for each package.

```

I believe it could be made to work, by running a series of updates.
but this just seems easier and cleaner.

isuruf commented:
Okay. I think your approach is easier then.

I'd rather this option was added in build.jl and the python version was decided at package build time. Now, the env variable is checked the first time a Conda call is made because miniconda is downloaded lazily.

@stevengj, what do you think?

stevengj commented:
I would recommend doing checking this in `deps/build.jl` (and storing the result in `deps.jl`) rather than every time the package is loaded.     You'll probably want to use `CONDA_JL_VERSION` in analogy with `CONDA_JL_HOME`, which we already have.  You'll want to save the environment variable in a file, e.g. like [IJulia does for JUPYTER](https://github.com/JuliaLang/IJulia.jl/blob/ad14a20472c7804a04178bb3be4865ac9b806748/deps/build.jl#L24), so that `Pkg.build` "remembers" its value if the user doesn't set the environment variable the next time the package is built.  And if the value of this variable is changed from the previous build, you'll want to re-do the miniconda install.

oxinabox commented:
Yes, these are good points.
I will make this so.

oxinabox commented:
Done.
I changed the Enviroment variable to CONDA_JL_VERSION,
but I left the variable name in the package MINICONDA_VERSION -- because that is what it is -- itis the version of the miniconda script used for installing.
(As compared to the version of the Conda.jl package.)

I added it to the deps file, rather than doing it every run.
I did a bit of a rewrite of the build.jl,
which you can see.
Because I can't just `include("deps.jl")` directly, if it exists, and one of the Environment variables is not defined -- since the other might be, and just including it would overwrite that value.

I think this code is cute,
but it might be "too clever"
the less cute way of doing it, is to use a different variable name in `build.jl` to in `deps.jl`

stevengj commented on README.md:
```diff
@@ -40,6 +40,15 @@ Basic package managing utilities are provided in the Conda module:
+This controls the initial version of python uses in the setup of Conda itself,
+and thus the default version of python for all dependancies.
+You can change the version by setting the enviroment variable `CONDA_JL_VERSION` to `"3"`, prior to installing the Conda.jl package.
+If you change this after Conda.jl has been added, you will need to rerun `Pkg.build("Conda")`.
```
In order for re-running `Pkg.build` to work, doesn't `build.jl` need to delete the existing install if it detects that the version is different from the installed version?

oxinabox replied:
Presumably, yes.
Thinking about this, 
Perhaps I should add more detail on it to the readme, and tell the user to do it.
With warnings.
Since deleting the ROOTENV, will break any packages they are using, that used Conda.jl to install their dependencies.
Which does not sound like something that should happen automatically.

stevengj replied:
The simplest thing might be for `Pkg.build` to throw an exception if it detects that a different version of conda has already been installed; the exception's error message can tell the user what directory they need to delete, and the problems it might cause.

oxinabox commented:
>Presumably, yes, (I've not tried running the miniconda2.sh and miniconda3.sh pointing at the same directory -- it might overwrite. Or do a soft upgrade. But I doubt it.)

I have confirms that this does *not* work.
And so have made the build.jl throw an error, as discussed.

stevengj commented on deps/build.jl:
```diff
@@ -12,7 +12,19 @@ end
-deps = """# Generated from $(@__FILE__) on $(now())
+if isdir(ROOTENV) && MINICONDA_VERSION != DefaultDeps.MINICONDA_VERSION
+    error("""Miniconda version changed, since last build.
+However, an Root Enviroment already exists at $(ROOTENV).
```
a root environment

stevengj commented on deps/build.jl:
```diff
@@ -12,7 +12,19 @@ end
+    error("""Miniconda version changed, since last build.
+However, an Root Enviroment already exists at $(ROOTENV).
+Setting Miniconda version is not supported for existing Root Enviroments.
+To leave Miniconda version as it is unset the CONDA_JL_VERSION enviroment variable, and rebuild.
```
comma after "as it is", no comma before "and" (also in the next sentence)

stevengj commented on deps/build.jl:
```diff
@@ -12,7 +12,19 @@ end
+To change Miniconda version, you must delete the Root Enviroment, and rebuild.
+WARNING: deleting the root enviroment will delete all the packages in it.
+This will break any Julia packages that have used Conda to install their dependancies.
+These will require rebuilding.
```
Not necessarily.  If a PyCall-based package used `pyimport_conda("foo")`, then the package will automatically re-install the next time that package is used.

Maybe, "Some Julia packages that use Conda to install their dependencies may need to be re-built."

oxinabox commented:
Cool,
would it be possible to have a new release tagged?
I want to use this to fix: https://github.com/malmaud/TensorFlow.jl/issues/251

isuruf commented on .travis.yml:
```diff
@@ -6,6 +6,10 @@ julia:
     - nightly
+env:
+    - MINICONDA_VERSION="2"
+    - MINICONDA_VERSION="3"
```
This should be `CONDA_JL_VERSION` right?

oxinabox replied:
Good catch.
I forgot to update that
I'll make a quick new PR

oxinabox replied:
https://github.com/JuliaPy/Conda.jl/pull/76

stevengj commented:
Tagged 0.6.

stevengj commented:
https://github.com/JuliaLang/METADATA.jl/pull/9901

GlenHertz commented:
Hi. Is it possible to select the Python sun-version. I need Python 3.4.x.

isuruf commented on deps/build.jl:
```diff
@@ -1,14 +1,33 @@
-    ROOTENV = abspath(dirname(@__FILE__), "usr")
+module DefaultDeps
+    if isfile("deps.jl")
+        include("deps.jl")
```
If this comes from a previous version, `MINICONDA_VERSION` will not be set and will throw an error. See https://github.com/JuliaPy/Conda.jl/issues/85

oxinabox replied:
darn I thought this was a clever solution -- but clearly did not think it fully through.
Still simple patch is possible
